### DISCUSSION TITLE
Contype for Knitro

### DISCUSSION THREAD

**Discussion by tinusd** [casadi/casadi#3260] [Q&A] [answered]
Hi there!

I realized that the contype option of Knitro has no effect. I then realized that the API of post Knitro 10.X no longer uses the cType array to specify the constraint types. Instead, you now have to use use dedicated functions to add linear, quadratic, or conic structures, see https://www.artelys.com/docs/knitro/3_referenceManual/callableLibraryAPI.html. I confirmed this with Knitro support: https://groups.google.com/g/KNITRO/c/ZJedNEzUiGU

Can I make a feature request for this? I don't mind if there will be some manual labour needed on my behalf, such as needing to use an option that is passed to CasADi's nlpsol where I have to give an array with the linear/quadratic coefficients (such as A and Q, where Ax and x'*Q*x).

Thank you, and kind regards

**Comment by jgillis** [comment, marked as answer]
Thanks for the warning, I had no idea about this API change..

**Reply by RenaudSlt** [reply]
Hello,

The latests versions of Casadi (>3.6.2) are compatible with Knitro 13.2 which, compared to Knitro 10.3, handles linear and quadratic expressions in dedicated functions. So there is room for improvement indeed.

I am curious about how Casadi handles linear and quadratic expressions and how it could be recovered by Knitro. I see some potentially interesting function in the C++ API, namely is_linear(), is_quadratic(), linear_coeff() and quadratic_coeff().

https://web.casadi.org/api/html/d9/dc2/group__expression__tools.html

Regards
